# =============================
# Imports & path
# =============================
import os
import sys
import grpc
import jwt
import time

# Ajouter grpc/ au PYTHONPATH pour trouver proto
ROOT_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.append(ROOT_DIR)

from proto import api_pb2, api_pb2_grpc

# =============================
# Configuration
# =============================
SERVER_ADDRESS = "localhost:50051"
SECRET_KEY = "raron"   # doit √™tre le m√™me que dans server.py

# =============================
# G√©n√©rer un JWT (simulation login)
# =============================
def generate_token(user_id=1):
    payload = {
        "user_id": user_id,
        "iat": int(time.time()),
        "exp": int(time.time()) + 3600  # 1 heure
    }
    token = jwt.encode(payload, SECRET_KEY, algorithm="HS256")
    return token

# =============================
# Client gRPC
# =============================
def main():
    print("üîå Connexion au serveur gRPC...")
    channel = grpc.insecure_channel(SERVER_ADDRESS)
    price_stub = api_pb2_grpc.PriceServiceStub(channel)
    user_stub  = api_pb2_grpc.UserServiceStub(channel)
    # G√©n√©rer token JWT
    token = generate_token(user_id=1)
    metadata = [("authorization", token)]

    # =============================
    # 1Ô∏è‚É£ Pr√©diction de prix
    # =============================
    try:
        print("üìä Envoi de la requ√™te PredictPrice...")
        request = api_pb2.PriceRequest(
            proprety_type="Apartment",
            surface=106.0,
            bedroom=2,
            bathroom =1,
            address="Rabat" ,
            city= "Rabat" ,
            principale= "Rabat-Sal√©-K√©nitra",
            
        )

        response = price_stub.PredictPrice(request, metadata=metadata)
        print("‚úÖ Prix pr√©dit :", response.price)

    except grpc.RpcError as e:
        print("‚ùå Erreur PredictPrice :", e.details())
        return

    # =============================
    # 2Ô∏è‚É£ Historique des pr√©dictions
    # =============================
    try:
        print("\nüìú R√©cup√©ration de l'historique...")
        history_request = api_pb2.UserHistoryRequest()

        history_response = user_stub.GetHistory(
            history_request,
            metadata=metadata
        )
        if not history_response.predictions:
            print("‚ÑπÔ∏è Aucun historique trouv√©")
        else:
            for h in history_response.predictions:
                print(
                    f"- {h.proprety_type} | "
                    f"- {h.surface} m¬≤ | "
                    f"{h.bedroom} chambres | "
                    f"{h.bathroom} WC | "
                    f"{h.address} Addresse  | "
                    f"{h.city} Ville | "
                    f"{h.principale} Region | "
                    f"Prix: {h.price} | "
                    f"Date: {h.created_at}"
                )

    except grpc.RpcError as e:
        print("‚ùå Erreur GetHistory :", e.details())

# =============================
# Main
# =============================
if __name__ == "__main__":
    main()
